Generate Timestamp & Duration schemas

bryfox opened the pull request:
### Changelog
None

### Docs
Docs PR forthcoming

### Description

This adds `Duration` and `Timestamp` to `FoxgloveMessageSchema` definitions (rather than treating them as primitives), so that we can begin documenting them as such in FG-11348.

Message schemas are now documented as part of the SDK (https://docs.foxglove.dev/docs/sdk/schemas), and this matches the "Timestamp" terminology used there.

In some cases, our types use the name "Time" rather than "Timestamp". That hasn't changed here; documentation will link to equivalent reference implementations.

For the most part, the generated schemas are unchanged. However:

- JSONSchemas now include Timestamp and Duration
- for FlatBuffers, I removed the documented `sec` range which was outside the range of int32.
- Some definitions now inherit the schema & field descriptions for the time-related types

gasmith commented on python/foxglove-sdk/python/foxglove/schemas/__init__.py:
```diff
@@ -100,8 +100,11 @@
 ]
 
+
+Time = Timestamp
```
I wonder whether we should do this the other way around. The `Timestamp` name feels like a protobuf-specific quirk, based on our decision to use that underlying well-known type for encoding.

amacneil replied:
I actually think `Timestamp` is probably a better name for it (worth doing an internal poll on it). E.g. would be nice to standardize on Timestamp everywhere in the schemas + SDK + app and not have an alias like this.

I'm not sure how practical it would be to update our flatbuffer and IDL schemas to use `Timestamp` instead of `Time` - the app could support both but it might cause some disruption.

amacneil replied:
@defunctzombie would it be possible to standardize our name here on `Timestamp` vs `Time`?

I guess in the app its currenty normalized to `Time`? Not sure what that affects tho.

defunctzombie replied:
Standardize in what sense? I don't think we lean too heavily on the string "Time" vs "Timestamp" in the code that handles message parsing and schemas - it is more prevalent in documentation and various type names.

Here's a few places where we reference "time" types in the docs:

- https://docs.foxglove.dev/docs/visualization/message-schemas/built-in-types#time
- https://docs.foxglove.dev/extension-api/type-aliases/Time
- https://docs.foxglove.dev/extension-api/type-aliases/MessageEvent#receivetime

It would be quite a lift to change the `Time` type to `Timestamp` in the various apis - likely we would alias and look at moving over piece by piece so `Time` would never fully go away but we could use `Timestamp` as the type in typescript.

gasmith replied:
> Some inconsistency is unavoidable since we want to use native types where possible (e.g., ROS; protobuf).

What's the value of using the native protobuf Timestamp and Duration types, which have different field names, different component types, and different ranges of representable values? It just seems like a source of confusion.

gasmith replied:
I'd like to challenge that a little.

If I'm writing with the foxglove SDK, and reading with some third-party tooling, I'd be confused by the fact that I logged my time in one format (the foxglove schema), and I'm reading it in a different format (the protobuf well-known types).

If I'm writing with third-party tooling, and reading with foxglove app, presumably the app already knows how to interpret either kind of timestamp. If not, it should. If we suppose reading MCAPs with the SDK as has been suggested, then I think it's the same argument as the paragraph above: I'd rather read exactly what I wrote, not a translation of it. Especially considering that foxglove timestamp/duration types are less expressive than the protobuf counterparts.

bryfox replied:
I've captured the above points (and out-of-band discussion) in FG-11450. This is specifically related to serialization considerations in the SDK, and need not block documentation of Foxglove schemas. I think the naming question (Timestamp vs Time) remains unanswered, but I'll start a new thread for that.

amacneil commented on typescript/schemas/src/internal/types.ts:
```diff
@@ -1,6 +1,7 @@
+  | "int32"
   | "uint32"
   | "boolean"
   | "bytes"
```
should we remove `time` and `duration` from this enum if they are now real schemas?

bryfox replied:
It's not clear to me what "primitive" means since it looks like pretty much everything here represents time as a sec/nsec pair; maybe it's more of an indication to use "built-in" types?

We could remove it here and adjust all the code to generate identical schemas, but since this is a public export, I'm inclined to leave it alone for now.

amacneil commented on typescript/schemas/src/internal/types.ts:
```diff
@@ -38,6 +39,7 @@ export type FoxgloveMessageSchema = {
   name: string;
   description: string;
   rosEquivalent?: keyof typeof import("@foxglove/rosmsg-msgs-common").ros1;
+  protoEquivalent?: "Timestamp" | "Duration";
```
would be better for the docs if we referred to them by the fully qualified name like `google.protobuf.Timestamp`

defunctzombie commented on typescript/schemas/src/internal/schemas.ts:
```diff
@@ -1,5 +1,45 @@
+  type: "message",
+  name: "Time",
+  description: "A timestamp composed of seconds and nanoseconds",
+  rosEquivalent: "std_msgs/Time",
```
How do we support ROS2 equivalent? which is `builtin_interfaces/Time` ?

bryfox replied:
Good question.

https://github.com/foxglove/foxglove-sdk/blob/aa7ab0727e054900642ffe2cc218bf684f3a0a53/typescript/schemas/src/internal/generateRos.ts#L32

Rather than duplicating those hardcoded values in our documentation too, I think I'll add a `ros2Equivalent`.

defunctzombie reviewed (approved):
👍 one open question on how we handled the `ros2 equivalent` behavior

bryfox commented:
We have one remaining question about naming — do we want to document the schema as `Time` or `Timestamp`? @amacneil [stated a preference for Timestamp](https://github.com/foxglove/foxglove-sdk/pull/391#discussion_r2071982760); an internal poll yielded no other votes. We use "Time" for documenting the extension API, and in user scripts. We use "Timestamp" in the SDK schema.

My reasoning for "Time" is in the description. I'll augment the documentation to describe some of the nuance here — naming, support for well-known types, differences with flatbuffers. I don't think an alias in the SDK (for discoverability) is a bad idea, but can remove that.

If we want to instead use "Timestamp", I'll close this PR and file an issue to first deal with that naming in the app where needed.

amacneil commented:
Copying my comment from slack

I think its worth renaming the schema to Timestamp and either updating the app or aliasing Time to Timestamp in the app.
this aligns better with industry terminology (timestamp == datetime, time == time sans date)

- [Protobuf Timestamp](https://protobuf.dev/reference/php/api-docs/Google/Protobuf/Timestamp.html)
- [Apache Arrow Timestamp](https://arrow.apache.org/docs/python/timestamps.html)
- [Postgres Timestamp](https://www.postgresql.org/docs/current/datatype-datetime.html)
- ros 2 calls the [schema Time and the field name ](https://docs.ros2.org/foxy/api/std_msgs/msg/Header.html)stamp - ros doing its own thing as always

bryfox commented:
I've updated the schema names to Timestamp. I kept the "Time" name in a couple of definitions to avoid a breaking change right now (see PR description), and to unblock the documentation work. I'll file follow-up issues for this.

gasmith commented on python/foxglove-sdk/python/foxglove/schemas/__init__.py:
```diff
@@ -111,6 +113,7 @@
     "CubePrimitive",
     "CylinderPrimitive",
     "Duration",
+    "Duration",
```
Duplicate

bryfox commented:
> For the sake of moving forward then lets go with Timestamp. App is still gonna need to accept both for backwards compatibility.

👍 I think this makes even more sense now given the new docs organization. I've brought this up-to-date with the main branch, and updated the description here. I have also updated a documentation PR to render previews for this proposed change.

jtbandes commented on typescript/schemas/src/internal/generateRos.ts:
```diff
@@ -163,7 +174,11 @@ export function generateRosMsgDefinition(
+        } else if (field.type.schema.name === "Duration") {
+          fieldType = "duration";
+        } else if (field.type.schema.rosEquivalent != undefined) {
           fieldType = field.type.schema.rosEquivalent;
```
seems weird to me that we aren't using rosEquivalent for this

jtbandes commented on typescript/schemas/src/internal/types.ts:
```diff
@@ -1,11 +1,4 @@
-  | "uint32"
-  | "boolean"
-  | "bytes"
-  | "time"
-  | "duration";
```
after seeing the impact of this elsewhere, it feels like we went from having special-case logic to deal with the fact that time/duration were almost-but-not-quite-primitives, to now having special-case logic to deal with the fact that they are almost-but-not-quite-schemas. I think my takeaway would be that having them be a separate `{ type: "time" }` (i.e. not lumping them in with all other `nested` types) might have resulted in less special casing. But none of this really matters too much as long as the right results are generated, because this api is all internal anyway. (hence how we never have cleaned up the `type: { type: ...` wart

bryfox commented on scripts/generate.ts:
```diff
@@ -162,16 +166,25 @@ async function main({ clean }: { clean: boolean }) {
   await logProgress("Generating FlatBuffer definitions", async () => {
     await fs.mkdir(path.join(outDir, "flatbuffer"), { recursive: true });
     await fs.writeFile(path.join(outDir, "flatbuffer", "ByteVector.fbs"), BYTE_VECTOR_FB);
+    // The name "Time" is kept for backwards compatibility.
     await fs.writeFile(path.join(outDir, "flatbuffer", "Time.fbs"), TIME_FB);
```
I'm assuming that flatbuffer codegen must produce different output if we change the names; they're effectively a new type that would break people's integrations.

I think that might be a step we want to take, but I would also like to avoid doing so in this PR.

The JSON schemas don't offer the old "Time" name. I added both Timestamp and Duration here, but we didn't output anything for those before.

bryfox commented on typescript/schemas/src/internal/generateJsonSchema.ts:
```diff
@@ -1,50 +1,63 @@
+      title: "duration",
+      properties: {
+        sec: { type: "integer" },
+        nsec: { type: "integer", minimum: 0, maximum: 999_999_999 },
```
It is unfortunate. I don't know why we used unsigned, but ROS is as good a guess as I could have.

> Hopefully at the end of the day the values are treated the same no matter where they came from?

I presume that's true for the app. I don't know how people are using these JSON schemas in particular, but restricting the range seems like a reasonable starting point, at least.

jtbandes commented on yarn.lock:
```diff
@@ -764,6 +764,13 @@ __metadata:
   languageName: node
   linkType: hard
 
+"@foxglove/message-definition@npm:^0.3.1":
```
Yeah maybe -- I think the challenge is that it makes dependabot PRs sometimes not merge cleanly, since dependabot sadly [doesn't run dedupe](https://github.com/dependabot/dependabot-core/issues/5830). I believe in some repos (maybe only internal repos?) we have a CI job that will run yarn dedupe on dependabot PRs.

jtbandes reviewed (approved):
Looks good!

It's probably not super critical, but the deduped yarn.lock still shows multiple versions of `@foxglove/rosmsg` and `@foxglove/message-definition` (i.e. new versions being added without the old ones being removed)

bryfox commented on typescript/schemas/src/internal/schemas.ts:
```diff
@@ -1,13 +1,55 @@
+    {
+      name: "nsec",
+      type: { type: "primitive", name: "uint32" },
+      description: "The number of nanoseconds in the positive direction",
```
I copied this from our existing docs. It's a little weird since we're not consistent about signed/unsigned; I'll file a follow-up issue.
